plex/player.py

```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#  Copyright (C) 2019 Sascha Montellese <montellese@kodi.tv>
+#
+#  SPDX-License-Identifier: GPL-2.0-or-later
+#  See LICENSES/README.md for more information.
+#
+import threading
+
+import xbmc
+import xbmcmediaimport
+
+from plex.api import Api
+from plex.constants import PLEX_PROTOCOL, PLEX_PLAYER_PLAYING, PLEX_PLAYER_PAUSED, PLEX_PLAYER_STOPPED
+from plex.server import Server
+
+from lib.utils import log, mediaProvider2str
+
+REPORTING_INTERVAL = 5 # seconds
+
+class Player(xbmc.Player):
+    def __init__(self):
+        '''Initializes the player'''
+        super(xbmc.Player, self).__init__()
+
+        self._providers = {}
+        self._monitor = xbmc.Monitor()
+        self._lock = threading.Lock()
+
+    def AddProvider(self, mediaProvider):
+        '''Adds a media provider to the player'''
+        if not mediaProvider:
+            raise ValueError('invalid mediaProvider')
+
+        self._providers[mediaProvider.getIdentifier()] = mediaProvider
+
+    def RemoveProvider(self, mediaProvider):
+        '''Removes the associated media provider'''
+        if not mediaProvider:
+            raise ValueError('invalid mediaProvider')
+
+        del self._providers[mediaProvider.getIdentifier()]
+
+    def onPlayBackStarted(self):
+        '''Triggered when xbmc.Player is started'''
+        self._reset()
+        self._file = self.getPlayingFile()
+
+    def onAVStarted(self):
+        '''Triggered when playback actually starts'''
+        self._startPlayback()
+        self._syncPlaybackState(time=0, state=PLEX_PLAYER_PLAYING)
+
+    def onPlayBackSeek(self, time, seekOffset):
+        '''Triggered when seeking.'''
+        self._syncPlaybackState(time=self._getPlayingTime())
+
+    def onPlayBackSeekChapter(self, chapter):
+        '''Triggered when seeking chapters.'''
+        self._syncPlaybackState(time=self._getPlayingTime())
+
+    def onPlayBackPaused(self):
+        '''Triggered when playback is paused.'''
+        self._syncPlaybackState(time=self._getPlayingTime(), state=PLEX_PLAYER_PAUSED)
+
+    def onPlayBackResumed(self):
+        '''Triggered when playback is resumed after a pause'''
+        self._syncPlaybackState(time=self._getPlayingTime(), state=PLEX_PLAYER_PLAYING)
+
+    def onPlayBackStopped(self):
+        '''Triggered when playback is stopped'''
+        self._syncPlaybackState(state=PLEX_PLAYER_STOPPED)
+        self.onPlayBackEnded()
+
+    def onPlayBackEnded(self):
+        '''Triggered when playback ends. Resets the player state and inherently kills the reporting loop'''
+        self._reset()
```

Comment:
Don't you need to report any specific "action"? Is not reporting any progress anymore enough?

Reply:
Not in case of plex. Stop already clears everything server side

Reply:
Where do you tell the Plex server that playback has stopped? IIRC the difference between `onPlayBackStopped()` and `onPlayBackEnded()` is that the former is triggered if the user manually stops playback and the latter is triggered if playback ends at the end of the video / file.

Reply:
Will add